### PR TITLE
Customization filters

### PR DESCRIPTION
**Pull request by Raruto**
As mentioned here: https://github.com/afragen/wp-dependency-installer/pull/37, I add the following filters to allow greater customization by third-party developers ( example of usage below ):

```php
// Additional filters.
add_filter( 'wp_dependency_notices', 'wpdi_notices', 10, 2 );  // show additional admin notices.
add_filter( 'wp_dependency_required_label', '__return_false' ); // disable "Required Plugin" label.
add_filter( 'wp_dependency_required_row_meta', '__return_false' );  // disable "Required by" row.
add_filter( 'wp_dependency_unset_action_links', '__return_false' ); // prevent removal of actions link.

// Prevent single deactivation of required plugins.
add_action( 'admin_action_deactivate', 'wpdi_before_deactivate_plugin' );

// Add required plugin asterisk.
add_action( 'after_plugin_row', 'wpdi_dependency_scripts' );

/**
 * Prevent admin ‘?action=deactive’ for required plugins.
 *
 * @return void
 */
function wpdi_before_deactivate_plugin() {
	if ( ! isset( $_REQUEST['plugin'] ) ) {
		return;
	}

	global $current_screen;

	if ( ! isset( $current_screen ) || ! isset( $current_screen->id ) || 'plugins' !== $current_screen->id ) {
		return;
	}

	$wpdi        = WP_Dependency_Installer::instance();
	$slug        = $_REQUEST['plugin'];
	$is_required = $wpdi->is_required( $slug );

	if ( $is_required ) {
		wp_safe_redirect( add_query_arg( 'wpdi_required', $slug, $_SERVER['HTTP_REFERER'] ) );
		exit();
	}

}

/**
 * Get plugin notices.
 *
 * @param string $slug Plugin slug.
 *
 * @return array Admin notices.
 */
function wpdi_notices( $notices, $slug ) {
	$wpdi = WP_Dependency_Installer::instance();

	if ( ! $wpdi->is_active( $slug ) ) {
		return $notices;
	}

	$dependency = $wpdi->get_config( $slug );

	foreach ( $dependency['sources'] as $source ) {
		// Check if we were trying to deactivate a manadatory plugin.
		if ( isset( $_REQUEST['wpdi_required'] ) && $slug === $_REQUEST['wpdi_required'] ) {
			array_unshift(
				$notices, [
					'status'  => 'error',
					/* translators: %s: Plugin name */
					'message' => sprintf( esc_html__( 'The %s plugin is a mandatory plugin.' ), $dependency['name'] ),
					'source'  => $source,
				]
			);
		}
	}

	return $notices;
}

/**
 * Add some custom styling.
 *
 * @param string $slug Plugin slug.
 *
 * @return void
 */
function wpdi_dependency_scripts( $plugin_file ) {
  if ( WP_Dependency_Installer::instance()->is_required( $plugin_file ) ) {
    print '<script>jQuery("tr[data-plugin=\'' . $plugin_file . '\'] .plugin-title > strong:first-child").append(\'<span class="wpdi-required"> *</span>\');</script>';
  }
}
```

### CODE REVIEW

**Comment by Raruto** [comment]
@afragen this pull adds the following filters:


```php
// Allow us to add additional admin notices.
add_filter( 'wp_dependency_notices', 'wpdi_notices', 10, 2 ); 

// Allow us to disable "Required Plugin" label ( plugin_actions_links )
add_filter( 'wp_dependency_required_label', '__return_false' );

// Allow us to disable disable "Required by" row ( plugin_row_meta ).
add_filter( 'wp_dependency_required_row_meta', '__return_false' );

// Allow us to prevent arbitrary removal of plugin actions link.
add_filter( 'wp_dependency_unset_action_links', '__return_false' );
```

and this is the output for the previous example:

customization-filters

I also rewrote the **`get_dependency_sources`** and **`get_dismiss_label`** functions just to have a more reusable code (I had already done it in the previous pull, but maybe you hadn't seen it ...).

**NB** if you give me some time, I also add some PHP Docs for each of these new filters...

File: wp-dependency-installer.php [hunk 1]
```diff
@@ -271,6 +271,8 @@ public function admin_init() {
 						$this->notices[] = $this->install_notice( $slug );
 					}
 				}
+
+				$this->notices = apply_filters( 'wp_dependency_notices', $this->notices, $slug );
```
Comment (Raruto) [inline wp-dependency-installer.php:286]: TODO: add PHP Docs here

File: wp-dependency-installer.php [hunk 2]
```diff
@@ -632,19 +628,27 @@ public function modify_plugin_row( $plugin_file ) {
 		 * @return mixed
 		 */
 		public function unset_action_links( $actions, $plugin_file ) {
-			if ( isset( $actions['delete'] ) ) {
-				unset( $actions['delete'] );
-			}
-			if ( isset( $actions['deactivate'] ) ) {
-				unset( $actions['deactivate'] );
+
+			if ( apply_filters( 'wp_dependency_unset_action_links', true ) ) {
```
Comment (Raruto) [inline wp-dependency-installer.php:650]: TODO: add PHP Docs here

File: wp-dependency-installer.php [hunk 3]
```diff
@@ -632,19 +628,27 @@ public function modify_plugin_row( $plugin_file ) {
...
+				}
+
+				if ( isset( $actions['deactivate'] ) ) {
+					unset( $actions['deactivate'] );
+				}
 			}
 
-			/* translators: %s: opening and closing span tags */
-			$actions = array_merge( [ 'required-plugin' => sprintf( esc_html__( '%1$sRequired Plugin%2$s' ), '<span class="network_active" style="font-variant-caps: small-caps;">', '</span>' ) ], $actions );
+			if ( apply_filters( 'wp_dependency_required_label', true ) ) {
```
Comment (Raruto) [inline wp-dependency-installer.php:667]: TODO: add PHP Docs here

File: wp-dependency-installer.php [hunk 4]
```diff
@@ -653,9 +657,13 @@ public function unset_action_links( $actions, $plugin_file ) {
 		 * @return void
 		 */
 		public function modify_plugin_row_elements( $plugin_file ) {
-			print '<script>jQuery(".inactive[data-plugin=\'' . $plugin_file . '\']").attr("class", "active");</script>';
-			print '<script>jQuery(".active[data-plugin=\'' . $plugin_file . '\'] .check-column input").remove();</script>';
-			print '<script>jQuery(".active[data-plugin=\'' . $plugin_file . '\'] .plugin-version-author-uri").append("<br><br><strong>' . esc_html__( 'Required by:' ) . '</strong> ' . $this->get_dependency_sources( $plugin_file ) . '");</script>';
+			print '<script>';
+			if ( apply_filters( 'wp_dependency_required_row_meta', true ) ) {
```
Comment (Raruto) [inline wp-dependency-installer.php:693]: TODO: add PHP Docs here

File: wp-dependency-installer.php [hunk 5]
```diff
@@ -666,21 +674,36 @@ public function modify_plugin_row_elements( $plugin_file ) {
 		 * @return string $dependents
 		 */
 		private function get_dependency_sources( $plugin_file ) {
```
Comment (Raruto) [inline wp-dependency-installer.php:708]: shorter and easier to read

File: wp-dependency-installer.php [hunk 6]
```diff
@@ -666,21 +674,36 @@ public function modify_plugin_row_elements( $plugin_file ) {
...
+		}
+
+		/**
+		 * Get formatted source string for text usage.
+		 *
+		 * @param  string $source plugin source.
+		 *
+		 * @return string friendly plugin name.
+		 */
+		private function get_dismiss_label( $source ) {
```
Comment (Raruto) [inline wp-dependency-installer.php:724]: code reusing ( also useful for third party developers )

File: wp-dependency-installer.php [hunk 7]
```diff
@@ -666,21 +703,36 @@ public function modify_plugin_row_elements( $plugin_file ) {
...
-				$sources,
-				function ( $value ) {
-					return ! empty( $value );
-				}
-			);
-			$dependents = implode( ', ', $sources );
-			$dependents = str_replace( '-', ' ', $dependents );
-			$dependents = ucwords( $dependents );
-			$dependents = str_ireplace( 'wp ', 'WP ', $dependents );
+			$sources = array_filter( $this->config[ $plugin_file ]['sources'] );
```
Comment (afragen) [inline wp-dependency-installer.php:710]: This is very nice.

**Review by afragen — APPROVED** [review]

**Comment by afragen** [comment]
I think the next version will be v3.0.0

**Comment by Raruto** [comment]
> I think the next version will be v3.0.0

👍

File: wp-dependency-installer.php (hunk 5 above)
Comment (Raruto) [reply wp-dependency-installer.php:708]: Could we find a more meaningful name? ( even if `private`, don't you think it could mislead? )

**Pull request merged**